python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+"""
+Classes for implementing keyed models.
+
+The use case that this addresses is where a client has a dataset with many keys - the distribution
+ of distinct keys amongst the is such that the total number of rows for every key value can be
+contained completely in memory on a single machine.
+
+This assumption is particularly enabling because clients may wish to apply more intricate single-
+machine models (such as a scikit-learn estimator) to every user.
+
+The API provided here generalizes the scikit-learn estimator interface to the Spark ML one; in
+particular, it allows clients to train their scikit-learn estimators in parallel over a grouped
+and aggregated dataframe.
+
+>>> import numpy as np
+>>> from sklearn.linear_model import LinearRegression
+>>> from pyspark.ml.linalg import Vectors
+>>> from pyspark.sql import SparkSession
+>>> from pyspark.sql.functions import udf
+>>> spark = SparkSession.builder.master("local").getOrCreate()
+>>> df = spark.createDataFrame([(user,
+...                              Vectors.dense([i, i ** 2, i ** 3]),
+...                              user + i + 2 * i ** 2 + 3 * i ** 3)
+...                             for user in range(3) for i in range(5)])
+>>> df = df.toDF("key", "features", "y")
+>>> km = KeyedEstimator(sklearnEstimator=LinearRegression(), yCol="y").fit(df)
+>>> printnp = lambda nparr: "[" + ", ".join("{:.2f}".format(x) for x in nparr) + "]"
+>>> coefs = udf(lambda lr: "intercept: {:.2f} coefs: {}".format(lr.intercept_, printnp(lr.coef_)))
+>>> km.keyedModels.select("key", coefs("estimator").alias("lr")).show(truncate=False)
++---+-----------------------------------------+
+|key|lr                                       |
++---+-----------------------------------------+
+|0  |intercept: 0.00 coefs: [1.00, 2.00, 3.00]|
+|1  |intercept: 1.00 coefs: [1.00, 2.00, 3.00]|
+|2  |intercept: 2.00 coefs: [1.00, 2.00, 3.00]|
++---+-----------------------------------------+
+<BLANKLINE>
+>>> input = spark.createDataFrame([(0, Vectors.dense(3, 1, -1))]).toDF("key", "features")
```

Comment:
Instead of inserting a blank line, I would move that to a subsquent paragraph with an explanation. "Now we have produced o model for each key, and we can apply it to test data:"